pkg/hostap/monitor.go

```diff
@@ -0,0 +1,239 @@
+	authenticatedMacs map[string]interface{}
+	mutex             sync.Mutex
+	stopWg            sync.WaitGroup
+	stop              chan interface{}
+	operState         netlink.LinkOperState
+}
+
+func (m *InterfaceMonitor) StartMonitor() error {
+	local, err := ioutil.TempFile(hostapdSocketDir, "hostapd_monitor_client")
+	if err != nil {
+		return err
+	}
+	os.Remove(local.Name())
+	conn, err := net.DialUnix(unixDgramProtocol, &net.UnixAddr{Name: local.Name(), Net: unixDgramProtocol},
+		&net.UnixAddr{Name: filepath.Join(hostapdSocketDir, m.IfName), Net: unixDgramProtocol})
+	if err != nil {
+		return err
+	}
+	m.hostApdConn = conn
+	m.authenticatedMacs = make(map[string]interface{})
+	m.stop = make(chan interface{})
+	m.stopWg.Add(3)
+	m.ifEventCh = make(chan netlink.LinkUpdate)
+	m.IfEventHandler.Subscribe(m.IfName, m.ifEventCh)
+	go m.handleHostapdReply()
+	go m.handleIfEvents()
+	go m.sendKeepAlive()
+	err = m.attachHostapd()
+	if err != nil {
+		m.StopMonitor()
+		return err
+	}
+	level.Info(m.Logger).Log("op", "monitor", "interface monitor started", m.IfName)
+	return nil
+}
+
+func (m *InterfaceMonitor) StopMonitor() {
+	close(m.stop)
+	m.IfEventHandler.Unsubscribe(m.IfName)
+	err := m.hostApdConn.Close()
+	if err != nil {
+		level.Error(m.Logger).Log("sockread", "error closing connection", m.IfName, err)
+	}
+	m.stopWg.Wait()
+}
+
+func (m *InterfaceMonitor) handleHostapdReply() {
+	defer m.stopWg.Done()
+	receivedByteArr := make([]byte, sockReadBufSize)
+	for {
+		select {
+		case <-m.stop:
+			return
+		default:
+			m.hostApdConn.SetReadDeadline(time.Now().Add(1 * time.Second))
+			size, err := m.hostApdConn.Read(receivedByteArr)
+			if err != nil {
+				if opErr, ok := err.(*net.OpError); ok && opErr.Timeout() {
+					continue
+				}
+				level.Error(m.Logger).Log("sockread", "error reading from connection", m.IfName, err)
+				return
+			}
+			if size == 0 {
+				continue
+			}
+			eventStr := string(receivedByteArr[:size])
+			err = m.handleHostapdEvent(eventStr)
+			if err != nil {
+				level.Error(m.Logger).Log("hostapd-event", eventStr, "interface", m.IfName, "error", err)
+			}
+		}
+	}
+}
+
+func (m *InterfaceMonitor) handleHostapdEvent(eventStr string) error {
+	eventStrSlice := strings.Split(eventStr, " ")
+	if len(eventStrSlice) < 2 {
+		return nil
+	}
+	eventKeyStr := eventStrSlice[0][strings.Index(eventStrSlice[0], ">")+1:]
+	switch eventKeyStr {
+	case staConnectedEvent, eapSuccessEvent:
+		return m.handleAutheticatedMac(eventStrSlice[1])
+	case staDisconnectedEvent, eapFailureEvent:
+		return m.handleDeAutheticatedMac(eventStrSlice[1])
+	default:
+		level.Info(m.Logger).Log("hostapd-event", "unhandled event", m.IfName, eventStr)
+	}
+	return nil
+}
+
+func (m *InterfaceMonitor) handleAutheticatedMac(macAddress string) error {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	m.authenticatedMacs[macAddress] = nil
+	err := trafficcontrol.AllowTrafficFromMac(m.IfName, macAddress)
+	if err != nil {
+		return err
+	}
+	return nil
```

Comment:
```suggestion
	return trafficcontrol.AllowTrafficFromMac(m.IfName, macAddress)
```
Isn't this the same, as long as appropriate logging already occurs in AllowTrafficFromMac?

Reply:
yes, done.